modules/utils/graph/Redif.cpp

```diff
@@ -0,0 +1,325 @@
+
+		blaze::DynamicMatrix<Tv> dist(n, n);
+		for (size_t i = 0; i < n; i++)
+			for (size_t j = 0; j < n; j++)
+			{
+				if (i == j)
+				{
+					dist(i, j) = 0;
+					continue;
+				}
+				blaze::DynamicVector<Tv, rowVector> rowi = row(dataSample, i);
+				blaze::DynamicVector<Tv, rowVector> rowj = row(dataSample, j);
+
+
+				std::vector<Tv> veci, vecj;
+				for (size_t k = 0; k < rowi.size(); ++k)
+				{
+					veci.push_back(rowi[k]);
+					vecj.push_back(rowj[k]);
+				}				
+
+				Metric metric;
+				dist(i, j) = metric(veci, vecj);
+			}
+		blaze::DynamicMatrix<size_t> knnMat(nNeighbors, n);
+		blaze::DynamicMatrix<Tv> knnDistMat(nNeighbors, n);
+
+		for (size_t i = 0; i < n; i++)
+		{
+			blaze::DynamicVector<size_t> tempIdx(n);
+			blaze::DynamicVector<Tv> tempDist(n);
+			for (size_t j = 0; j < n; j++)
+			{
+				tempIdx[j] = j;
+				tempDist[j] = dist(j, i);
+			}
+			std::cout << i << "-th" << std::endl;
+			Quicksort(tempIdx, tempDist, 0, n - 1);
+			//if (i == 58)
+			//{
+			//	std::cout << std::endl;
+			//	for (size_t j = 0; j < 30; j++)
+			//		std::cout << tempDist[j] << std::endl;
+			//}
+			for (size_t j = 0; j < nNeighbors; j++)
+			{
+				knnMat(j, i) = tempIdx[j + 1];
+				knnDistMat(j, i) = tempDist[j + 1];
+			}
+		}
+
+		blaze::DynamicMatrix<Tv> localDist(n, n);
+
+		for (size_t i = 0; i < nNeighbors; i++)
+			for (size_t j = 0; j < n; j++)
+				localDist(ixx(i, j), knnMat(i, j)) = knnDistMat(i, j);
+
+		blaze::DynamicMatrix<Tv> retLocalDist(n, n);
+		for (size_t i = 0; i < n; i++)
+			for (size_t j = 0; j < n; j++)
+				retLocalDist(i, j) = (localDist(i, j) > localDist(j, i)) ? localDist(i, j) : localDist(j, i);
+
+		return retLocalDist;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::calcWeightedGraphLaplacian(blaze::DynamicMatrix<Tv> localDist)
+	{
+		size_t n = localDist.rows();
+
+		/// Degree matrix
+		blaze::DynamicMatrix<Tv> D(n, n);
+		for (size_t i = 0; i < D.rows(); i++)
+		{
+			double sum = 0;
+			for (size_t j = 0; j < n; j++)
+				sum += localDist(i, j);
+			if (sum == 0) sum = 1;
+			D(i, i) = 1 / sum;
+		}
+
+		blaze::DynamicMatrix<Tv> A(localDist.rows(), localDist.columns());
+		A = D * localDist * D / n;
+
+		blaze::DynamicMatrix<Tv> D1(n, n);
+		for (size_t i = 0; i < D.rows(); i++)
+		{
+			double sum = 0;
+			for (size_t j = 0; j < n; j++)
+				sum += A(i, j);
+			if (sum == 0) sum = 1 / n;
+			D1(i, i) = sum;
+		}
+
+		blaze::DynamicMatrix<Tv> graphLap(n, n);
+		graphLap = D1 - A;
+
+		return graphLap;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::encode(blaze::DynamicMatrix<Tv> x, blaze::DynamicVector<Tv> & l_idx)
+	{
+		size_t nTrain = xTrain.rows();
+		size_t nX = x.rows();
+		size_t nIter = LArray.size();
+		/// Find nearest Neighbors of each record of x in xTrain
+		size_t l = 1; /// Number of nearest neighbors
+		l_idx = blaze::DynamicVector<Tv>(nX);

```

Comment:
if we discard l_idx content maybe it make sense return it as part of tuple<DynamicMatrix, DynamicVector> ? if we need to follow API, then please ignore my comment